# Project DataQuest 01

This is the first of several self guided projects from [DataQuest](https://dataquest.io). The goal of this project to show what type of apps are likely to attract more users on Google Play and the App Store.

* [Dataset for Google Play Store](https://www.kaggle.com/datasets/lava18/google-play-store-apps)
* [Dataset for Apple Store](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps)

In [43]:
!wget -O googleplaystore.csv https://dq-content.s3.amazonaws.com/350/googleplaystore.csv
!wget -O AppleStore.csv https://dq-content.s3.amazonaws.com/350/AppleStore.csv

563.90s - pydevd: Sending message related to process being replaced timed-out after 5 seconds


--2024-08-05 12:28:51--  https://dq-content.s3.amazonaws.com/350/googleplaystore.csv
Resolving dq-content.s3.amazonaws.com (dq-content.s3.amazonaws.com)... 54.231.231.81, 54.231.225.89, 54.231.138.217, ...
Connecting to dq-content.s3.amazonaws.com (dq-content.s3.amazonaws.com)|54.231.231.81|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1349314 (1.3M) [text/csv]
Saving to: ‘googleplaystore.csv’

googleplaystore.csv 100%[===================>]   1.29M  4.05MB/s    in 0.3s    

2024-08-05 12:28:51 (4.05 MB/s) - ‘googleplaystore.csv’ saved [1349314/1349314]



569.97s - pydevd: Sending message related to process being replaced timed-out after 5 seconds


--2024-08-05 12:28:57--  https://dq-content.s3.amazonaws.com/350/AppleStore.csv
Resolving dq-content.s3.amazonaws.com (dq-content.s3.amazonaws.com)... 52.216.251.52, 54.231.225.89, 54.231.138.217, ...
Connecting to dq-content.s3.amazonaws.com (dq-content.s3.amazonaws.com)|52.216.251.52|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 725761 (709K) [text/csv]
Saving to: ‘AppleStore.csv’

AppleStore.csv      100%[===================>] 708.75K  2.50MB/s    in 0.3s    

2024-08-05 12:28:57 (2.50 MB/s) - ‘AppleStore.csv’ saved [725761/725761]



In [44]:
def explore_data(dataset, start=0, end=1, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')  # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [45]:
play_store_buffer = open('googleplaystore.csv')
app_store_buffer = open('AppleStore.csv')

from csv import reader

play_store_raw = reader(play_store_buffer)
app_store_raw = reader(app_store_buffer)

play_store_ds = list(play_store_raw)
app_store_ds = list(app_store_raw)

In [46]:
explore_data(play_store_ds, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10842
Number of columns: 13


Columns that currently exist for play_store dataset:

| Column Name      | Description                                                           |
| ---------------- | --------------------------------------------------------------------- |
| App              | The name of the app.                                                  |
| Category         | The category to which the app belongs.                                |
| Rating           | The average user rating of the app out of 5.                          |
| Reviews          | The total number of user reviews.                                     |
| Size             | The size of the app.                                                  |
| Installs         | The total number of times the app has been installed.                 |
| Type             | Indicates whether the app is free or paid.                            |
| Price            | The cost of the app (usually 0 for free apps).                        |
| Content Rating   | The age rating for the app (e.g., Everyone, Teen, Mature 17+).        |
| Genres           | The genres associated with the app.                                   |
| Last Updated     | The date when the app was last updated.                               |
| Current Ver      | The current version of the app.                                       |
| Android Ver      | The minimum Android version required to run the app.                  |


In [47]:
explore_data(app_store_ds, 0, 4, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7198
Number of columns: 16


Columns that currently exist for app_store dataset:

| Column Name        | Description                                                                 |
| ------------------ | --------------------------------------------------------------------------- |
| id                 | Unique identifier for the app.                                              |
| track_name         | The name of the app.                                                        |
| size_bytes         | The size of the app in bytes.                                               |
| currency           | The currency in which the price is listed.                                  |
| price              | The cost of the app.                                                        |
| rating_count_tot   | The total number of user ratings.                                           |
| rating_count_ver   | The number of user ratings for the current version.                         |
| user_rating        | The average user rating out of 5.                                           |
| user_rating_ver    | The average user rating for the current version out of 5.                   |
| ver                | The current version of the app.                                             |
| cont_rating        | The age rating for the app (e.g., 4+, 9+, 12+).                             |
| prime_genre        | The main category of the app.                                               |
| sup_devices.num    | The number of devices the app supports.                                     |
| ipadSc_urls.num    | The number of iPad screenshots available.                                   |
| lang.num           | The number of languages the app supports.                                   |
| vpp_lic            | Indicates if the app is available for Volume Purchase Program (VPP) licensing. |


## Cleaning play store data

First thing we need to do is clean the play store and app store of it's:
- duplicates
- non-free
- non-english
- incompatible data

First we look for "lengths", meaning rows that do not match the header length. This is corrupt data. Once we clean it, we print the length of play_store_ds. Once we see that the row count went down by one, we can move on to cleaning duplicates. Duplicates of App Name constitute bad data. We remove them destructively in favor of the first instance of the name in our dataset. Next, we isolate free apps from this dataset, also destructive against non-free apps. We look for text `Free` for applicable rows. Finally, we loop through text and, if we find non-English characters, we delete said row. The next set of code will illustrate this.

We repeat the process for the App Store, which should be cleaned in the same way. What we see is a drastic reduction in rows, with playstore starting at 10842 records and appstore starting at 7197 records with both decreasing to 4783 and 2263 records respectively.

In [48]:
def clean_lengths(ds):
    """Finds mismatched row lenghts. This is a destructive action!
    """
    for idx in range(len(ds)):
        if 0 < idx < len(ds):
            if len(ds[0]) != len(ds[idx]):
                del ds[idx]

def clean_dupes(ds, dupe_idx):
    """Cleans up rows with duplicate value. This is a destructive action!
    """
    seen = set()
    for idx in range(len(ds)):
        if 0 < idx < len(ds):
            dupe = ds[idx][dupe_idx]
            if dupe in seen:
                del ds[idx]
            else:
                seen.add(dupe)

def clean_non_free(ds, price_idx):
    """Finds non-free items based on price index provided. This is a destructive action!
    """
    # Given a price index, remove non-free items
    for idx in range(len(ds)):
        if 0 < idx < len(ds):
            price = ds[idx][price_idx]
            if price in ('0', '0.0', 'Free'):
                del ds[idx]

def clean_non_english(ds, text_idx):
    """Finds non-english items based on a textual index provided. This is a destructive action!
    """
    for idx in range(len(ds)):
        if 0 < idx < len(ds):
            text = ds[idx][text_idx]
            for t in text:
                if(ord(t)) > 127:
                    try:
                        del ds[idx]
                    except:
                        pass

In [49]:
# We clean the playstore data
explore_data(play_store_ds, 0, 0, True)
clean_lengths(play_store_ds)
print("Cleaning lengths:")
explore_data(play_store_ds, 0, 0, True)
print("Cleaning dupes:")
clean_dupes(play_store_ds, 0)
explore_data(play_store_ds, 0, 0, True)
print("Cleaning non-free:")
clean_non_free(play_store_ds, 6)
explore_data(play_store_ds, 0, 0, True)
print("Cleaning non-english:")
clean_non_english(play_store_ds, 0)
explore_data(play_store_ds, 0, 0, True)

Number of rows: 10842
Number of columns: 13
Cleaning lengths:
Number of rows: 10841
Number of columns: 13
Cleaning dupes:
Number of rows: 10033
Number of columns: 13
Cleaning non-free:
Number of rows: 5252
Number of columns: 13
Cleaning non-english:
Number of rows: 4783
Number of columns: 13


In [51]:
# We clean the playstore data
explore_data(app_store_ds, 0, 0, True)
print("Cleaning lengths:")
clean_lengths(app_store_ds)
explore_data(app_store_ds, 0, 0, True)
print("Cleaning dupes:")
clean_dupes(app_store_ds, 1)
explore_data(app_store_ds, 0, 0, True)
print("Cleaning non-free:")
clean_non_free(app_store_ds, 4)
explore_data(app_store_ds, 0, 0, True)
print("Cleaning non-english:")
clean_non_english(app_store_ds, 1)
explore_data(app_store_ds, 0, 0, True)

Number of rows: 3018
Number of columns: 16
Cleaning lengths:
Number of rows: 3018
Number of columns: 16
Cleaning dupes:
Number of rows: 3018
Number of columns: 16
Cleaning non-free:
Number of rows: 2342
Number of columns: 16
Cleaning non-english:
Number of rows: 2263
Number of columns: 16
